# Information Theory and the Game Wordle (Part 2 of 2)

## Goal of this lecture (Part 1 of 2):

Now that we know how the game Wordle works, and a basic understanding of Information Theory, we will apply it to find the optimal first guess for Wordle!

## Setting up our infrastructure

I've uploaded a file called Wordle.py that includes all of the possible guesses that Wordle will accept. This is the same as the one used in the Julia portion.

We are going to need to write some code to help make this process easier. First, we need a way to represent the information given to us for each guess. Since each box can be gray, yellow, or green, there is a total of $3^5 = 243$ kinds of outcomes that each word can give us.

In [8]:
import enum
import itertools
import Wordle
import math

class Results(enum.Enum):
    gray = 0
    yellow = 1
    green = 2

possibleInputs = Wordle.words + Wordle.answers # words and answers are disjoint, so combining them gives us all possible inputs
outcomes = tuple(itertools.product(Results, repeat=5)) # All tuples of length 5 where the values are Results.gray, Results.yellow, or Results.green

print("Num outcomes: " + str(len(outcomes))) # This tells us that we have all 243 possible outcomes
print("Num possibleInputs: " + str(len(possibleInputs)))

Num outcomes: 243
Num possibleInputs: 12972


Now we want a function that tells us the outcome of the guess compared to the answer. We'll also create a function that returns the number of times each possible outcome occured so that we can calculate probability later.

In [9]:
def getOutcome(guess, answer):
    outcome = [Results.gray, Results.gray, Results.gray, Results.gray, Results.gray]
    
    for i in range(0, 5):
        if guess[i] == answer[i]:
            outcome[i] = Results.green
        elif guess[i] in answer:
            outcome[i] = Results.yellow
    
    return tuple(outcome)

def getOutcomeDict(guess, wordList):
    # A diction that contains the number of each time an outcome appears
    outcomeDict = {}
    for outcome in outcomes:
        outcomeDict[outcome] = 0
        
    for answer in wordList:
        outcomeDict[getOutcome(guess, answer)] += 1
        
    return outcomeDict

print(getOutcome("crate", "crane"))

(<Results.green: 2>, <Results.green: 2>, <Results.green: 2>, <Results.gray: 0>, <Results.green: 2>)


Now, we want to create a function that returns the entropy/expected information that we would get for each word and a list of words. To do so, we will define a function that takes in a word. With that word, we will compare it to every other word and see what the outcome is. Then we can calculate the probabilities with these numbers.

Notice that, instead of just always using every Wordle word, I allow you to pass in a word list so that you can remove words from the list and get a different answer. This will allow use to find the changes in entropy for after subsequent guesses.

In [10]:
def entropy(guess, wordList):
    # A dictionary that contains the number of each time an outcome appears
    outcomeDict = getOutcomeDict(guess, wordList)
    totalOutcomes = len(wordList)
    
    expected = 0
    for outcome in outcomes:
        p = outcomeDict[outcome]/totalOutcomes #calculating probability
        if p != 0:
            expected += p * math.log(1/p, 2) #Entropy equation
            
    return expected

print(entropy("abyss", possibleInputs))

3.9687406244872463


Based on what value our entropy function gives us for a word, we can expect that word to divide our possible guesses in half that many number of times! Remember, an entropy value of 4 means that we cut our possibilities in half 4 times!

Now that we have a function that returns the entropy of each word. We can find the word with the highest entropy value by iterating through every single word and getting their entropy!

In [11]:
def getHighestEntropyIndex(wordList):
    highestEntropyIndex = 0
    highestEntropy = 0
    for i in range(len(wordList)):
        curEntropy = entropy(wordList[i], wordList)
        if curEntropy > highestEntropy:
            highestEntropy = curEntropy
            highestEntropyIndex = i
    return highestEntropyIndex
        
highestEntropyIndex = getHighestEntropyIndex(Wordle.answers)

In [12]:
print("The word with the highest entropy is at index " + str(highestEntropyIndex))
print("The word with the highest entropy is " + Wordle.answers[highestEntropyIndex])
print("The word's entropy is " + str(entropy(Wordle.answers[highestEntropyIndex], possibleInputs)))

The word with the highest entropy is at index 1668
The word with the highest entropy is raise
The word's entropy is 5.9197368425161985


The result of this code tells us that the word with the highest entropy is "raise" with an entropy value of almost 6. This means that, on average, guessing saner will cut down all the possible remaining guesses in half 6 times! So, that means that if we have 12,972 possible words at the beginning, there will only be about 202 possible guesses left!

Now, if you have watched the video by 3Brown1Blue, you might have noticed that my word is different from his. That is because of a couple of reasons:
1. He added code to consider other factors such as word frequency
2. He actually looks deeper into the search to see if how he could optimize later guesses, not just the first guess
3. He considered all possible guesses, I only considered all possible answers

If you were paying close attention, you might have noticed that I used "Wordle.answers" this time instead of the usual "possibleInputs". That's because "Wordle.answers" is the subset of all words that could possibly be answers. This makes the code run faster and it actually gives you a more accurate answer! The whole list of words is too long for this code to run on CoCalc.

So, now that we can find the word with the highest entropy, we need a function that will return us back the list of remaining words!

In [13]:
def getRemainingWordsListWithOutcome(guess, wordList, outcome):
    newWordList = []
    for word in wordList:
        add = True
        for i in range(0, 5):
            if outcome[i] == Results.green and word[i] != guess[i]:
                add = False
                continue
            elif outcome[i] == Results.yellow and guess[i] not in word:
                add = False
                continue
        if add:    
            newWordList.append(word)
    return newWordList

print(getRemainingWordsListWithOutcome("crate", possibleInputs, (Results.green, Results.green, Results.green, Results.gray, Results.green)))

['crake', 'crame', 'crape', 'crare', 'crate', 'craze', 'crave', 'crane']


Now that we can find the word with the highest Entropy and we can get the remaining words left after a guess, we can actually write a function that solves the game for us! I won't actually do that because it requires me to add user input in CoCalc, and I am not sure how to do that. But the code would work something like this:
1. Get the first optimal word (Which we know is saner)
2. Input saner into Wordle and get outcome
3. Get remaining words list with the outcome
4. Then get the next optimal word from the remaining list
5. Go back to step 2

# Conclusion
With a little bit of math and coding, we have written a program that can find us the word with the highest entropy so that, on average, it reduces the amount of possible guesses as much as possible. However, I would like to point out that this program can be further improved to give even better guesses. First, we can also give higher weights to words that are used more frequently, and words that have more common letters.

The program is also quite slow. The Youtuber 3Blue1Brown made a video where he showed that an optimization he tried created errors in the results. So I wanted to avoid those issues by making the code do all the work without trying to skip things.

However, in the end, it works! And it probably would work a lot better if it we running on my local machine than on CoCalc. But I definitely enjoyed making this and I am glad that it works. Now I am going to start all my games with the word Saner.

In [14]:
print("Goodbye, world!")

Goodbye, world!
